frontend/src/app/main/data/workspace/drawing/path.cljs

```diff
@@ -0,0 +1,697 @@
+            drag-events (->> ms/mouse-position
+                             (rx/take-until mouse-up)
+                             (rx/map #(drag-handler %)))]
+
+        (rx/concat (rx/of (add-node position))
+                   (rx/of (start-drag-handler))
+                   drag-events
+                   (rx/of (finish-drag))))
+      )))
+
+;; EVENT STREAMS
+
+(defn make-click-stream
+  [stream down-event]
+  (->> stream
+       (rx/filter ms/mouse-click?)
+       (rx/debounce 200)
+       (rx/first)
+       (rx/map #(add-node down-event))))
+
+(defn make-drag-stream
+  [stream down-event]
+  (let [mouse-up    (->> stream (rx/filter #(or (end-path-event? %)
+                                                (ms/mouse-up? %))))
+        drag-events (->> ms/mouse-position
+                         (rx/take-until mouse-up)
+                         (rx/map #(drag-handler %)))]
+    (->> (rx/timer 400)
+         (rx/merge-map #(rx/concat
+                         (rx/of (add-node down-event))
+                         (rx/of (start-drag-handler))
+                         drag-events
+                         (rx/of (finish-drag)))))))
+
+(defn make-dbl-click-stream
+  [stream down-event]
+  (->> stream
+       (rx/filter ms/mouse-double-click?)
+       (rx/first)
+       (rx/merge-map
+        #(rx/of (add-node down-event)
+                ::end-path))))
+
+(defn make-node-events-stream
+  [stream]
+  (->> (rx/merge
+        (->> stream (rx/filter (ptk/type? ::close-path)))
+        (->> stream (rx/filter (ptk/type? ::close-path-drag-start))))
+       (rx/take 1)
+       (rx/merge-map #(rx/empty))))
+
+;; MAIN ENTRIES
+
+(defn handle-drawing-path
+  [id]
+  (ptk/reify ::handle-drawing-path
+    ptk/UpdateEvent
+    (update [_ state]
+      (let [id (get-path-id state)]
+        (-> state
+            (assoc-in [:workspace-local :edit-path id :edit-mode] :draw))))
+
+    ptk/WatchEvent
+    (watch [_ state stream]
+      (let [mouse-down    (->> stream (rx/filter ms/mouse-down?))
+            end-path-events (->> stream (rx/filter end-path-event?))
+
+            ;; Mouse move preview
+            mousemove-events
+            (->> ms/mouse-position
+                 (rx/take-until end-path-events)
+                 (rx/throttle 50)
+                 (rx/map #(preview-next-point %)))
+
+            ;; From mouse down we can have: click, drag and double click
+            mousedown-events
+            (->> mouse-down
+                 (rx/take-until end-path-events)
+                 (rx/throttle 50)
+                 (rx/with-latest merge ms/mouse-position)
+
+                 ;; We change to the stream that emits the first event
+                 (rx/switch-map
+                  #(rx/race (make-node-events-stream stream)
+                            (make-click-stream stream %)
+                            (make-drag-stream stream %)
+                            (make-dbl-click-stream stream %))))]
+
+        (->> (rx/concat
```

Comment:
The `->>` is not necessary here i think.

Reply:
You're right. That's some code I forgot to remove :smile: